File: 1731-even-odd-tree/even-odd-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isEvenOddTree(TreeNode root) {
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        int level = 0;
        while (!q.isEmpty()) {
            int level_size = q.size();
            boolean is_odd_level = level % 2 == 1;
            int prev = -1;
            if (is_odd_level) {
                prev = Integer.MAX_VALUE;
            } else {
                prev = 0;
            }
            for (int i = 0; i < level_size; i++) {
                TreeNode cur = q.poll();
                if (is_odd_level) {
                    if (cur.val % 2 == 1 || prev <= cur.val) {
                        return false;
                    }
                } else {
                    if (cur.val % 2 == 0 || prev >= cur.val) {
                        return false;
                    }
                }
                if (cur.left != null) {
                    q.add(cur.left);
                }
                if (cur.right != null) {
                    q.add(cur.right);
                }
                prev = cur.val;
            }
            level++;
        }
        return true;
    }
}

File: 1731-even-odd-tree/even-odd-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isEvenOddTree(self, root: Optional[TreeNode]) -> bool:
        q = deque([root])
        level = 0
        while q:
            level_size = len(q)
            odd_lvl = level % 2 
            prev = float('inf') if odd_lvl else 0
            # print(lvl_num, prev)
            for _ in range(level_size):
                cur = q.popleft()
                if odd_lvl:
                    if cur.val % 2 == 1 or prev <= cur.val:
                        return False
                else:       # even lvl
                    if cur.val % 2 == 0 or prev >= cur.val:
                        return False
                if cur.left: 
                    q.append(cur.left)
                if cur.right: 
                    q.append(cur.right)
                prev = cur.val
            level += 1
        return True